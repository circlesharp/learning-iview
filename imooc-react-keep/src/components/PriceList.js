import React from 'react';
import PropTypes from 'prop-types';
import Ionicon from 'react-ionicons';

const PriceList = ({ items, onModifyItem, onDeleteItem }) => {
  return (
    <ul className="list-group list-group-flush">
      {
        items.map(item => (
          <li
            className="list-group-item d-flex justify-content-between align-items-center"
            key={item.id}
          >
            <span className="col-1 badge">
              <Ionicon
                className="rounded-circle"
                fontSize="30px"
                color="#fff"
                icon={item.category.iconName}
                style={{ backgroundColor: '#007bff', padding: '5px' }}
              />
            </span>
            <span className="col-5">
              {item.title}
            </span>
            <span className="col-2 font-weight-bold">
              {item.category.type === 'income' ? '+' : '-'}
              {item.price}元
            </span>
            <span className="col-2">{item.date}</span>

            <span
              className="icon-modify col-1"
              onClick={() => {onModifyItem(item)}}
            >
              <Ionicon
                className="rounded-circle"
                fontSize="30px"
                color="#fff"
                icon="ios-create-outline"
                style={{ backgroundColor: '#28a745', padding: '5px' }}
              />
            </span>
            <span
              className="icon-delete col-1"
              onClick={() => {onDeleteItem(item)}}
            >
              <Ionicon
                className="rounded-circle"
                fontSize="30px"
                color="#fff"
                icon="ios-close"
                style={{ backgroundColor: '#dc3545', padding: '5px' }}
              />
            </span>
          </li>
        ))
      }
    </ul>
  );
};

/* 使用 PropTypes 进行 props 检查 */
/* https://reactjs.org/docs/typechecking-with-proptypes.html#gatsby-focus-wrapper */
PriceList.propTypes = {
  items: PropTypes.array.isRequired,
  onModifyItem: PropTypes.func.isRequired,
  onDeleteItem: PropTypes.func.isRequired,
};

/* default props */
/* 与 vue 有点不一样，默认和必填不冲突 */
PriceList.defaultProps = {
  onDeleteItem: item => {console.error(item)},
};

export default PriceList;
