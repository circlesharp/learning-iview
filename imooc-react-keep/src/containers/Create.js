import React from 'react';
import { withRouter } from 'react-router-dom';
import withContext from '../WithContext';
import PriceForm from '../components/PriceForm';
import CategorySelect from '../components/CategorySelect';
import { Tabs, Tab } from '../components/Tabs';
import { TYPE_INCOME, TYPE_OUTCOME } from '../utility';

const tabsText = [TYPE_OUTCOME, TYPE_INCOME];

class Create extends React.Component {
  constructor(props) {
    super(props);

    const { items, categories } = this.props.data;
    const { id } = this.props.match.params;

    this.state = {
      selectedTab: (id && items[id]) ? categories[items[id].cid].type : TYPE_OUTCOME,
      selectedCategory: (id && items[id]) ? categories[items[id].cid] : null,
    };
  }

  async componentDidMount() {
    const { id } = this.props.match.params;
    const { editItem, categories } = await this.props.actions.getEditData(id);
    this.setState({
      selectedTab: (id && editItem) ? categories[editItem.cid].type : TYPE_OUTCOME,
      selectedCategory: (id && editItem) ? categories[editItem.cid] : null,
    });
  }

  onTabChange = index => {
    this.setState({ selectedTab: tabsText[index] });
  };

  selectCategory = category => {
    this.setState({ selectedCategory: category });
  };

  onCancelSubmit = () => {
    this.props.history.push('/');
  };

  submitForm = async (data, isEditMode) => {
    if (this.state.selectedCategory == null) {
      alert('please select category');
      return;
    }

    if (!isEditMode) {
      // create
      await this.props.actions.createItem(data, this.state.selectedCategory.id);
    } else {
      // update
      await this.props.actions.updateItem(data, this.state.selectedCategory.id);
    }

    this.props.history.push('/');
  };

  render() {
    const { items, categories } = this.props.data;
    const { selectedTab, selectedCategory } = this.state;
    const { id } = this.props.match.params;
    const editItem = (id && items[id]) ? items[id] : {};

    const filterCategories = Object.keys(categories)
      .filter(id => categories[id].type === selectedTab)
      .map(id => categories[id]);

    return (
      <div
        className="create-page py-3 px-3 rounded mt-3"
        style={{ background: '#fff' }}
      >
        {/* 1 Tabs */}
        <Tabs
          activeIndex={tabsText.indexOf(selectedTab)} // 可用 findIndex 代替
          onTabChange={this.onTabChange}
        >
          <Tab>支出</Tab>
          <Tab>收入</Tab>
        </Tabs>

        {/* 2 CategorySelect */}
        <CategorySelect
          categories={filterCategories}
          selectedCategory={selectedCategory}
          onSelectCategory={this.selectCategory}
        />

        {/* 3 PriceForm */}
        <PriceForm
          item={editItem}
          onFormSubmit={this.submitForm}
          onCancelSubmit={this.onCancelSubmit}
        />
      </div>
    );
  }
}

export default withRouter(withContext(Create));
