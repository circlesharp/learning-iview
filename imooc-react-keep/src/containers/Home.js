import React from 'react';
import { withRouter } from 'react-router-dom';
import Ionicon from 'react-ionicons';

import withContext from '../WithContext';
import PriceList from '../components/PriceList';
import MonthPicker from '../components/MonthPicker';
import TotalPrice from '../components/TotalPrice';
import CreateBtn from '../components/CreateBtn';
import Loader from '../components/Loader';
import PieChart from '../components/PieChart';
import { Tabs, Tab } from '../components/Tabs';

import { LIST_VIEW, CHART_VIEW, TYPE_INCOME, TYPE_OUTCOME, generateChartDataByCategory } from '../utility';

export const newItem = {
  id: '4',
  title: 'new title',
  price: 10,
  date: '2021-01-26',
  cid: '2',
};

const tabsText = [LIST_VIEW, CHART_VIEW];

export class Home extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      tabView: tabsText[0],
    };
  }

  componentDidMount() {
    this.props.actions.getInitialData();
  }

  changeView = index => {
    this.setState({ tabView: tabsText[index] });
  };
  changeDate = (year, month) => {
    this.props.actions.selectNewMonth(year, month);
  };
  modifyItem = item => {
    this.props.history.push(`/edit/${item.id}`);
  };
  createItem = () => {
    this.props.history.push('/create');
  };
  deleteItem = item => {
    this.props.actions.deleteItem(item);
  };

  render() {
    const { items, categories, currentDate, isLoading } = this.props.data;
    const { tabView } = this.state;
    const itemsWithCategory = Object.keys(items).map(id => {
      items[id].category = categories[items[id].cid];
      return items[id];
    });

    const chartOutcomeDateByCategory = generateChartDataByCategory(itemsWithCategory, TYPE_OUTCOME);
    const chartIncomeDateByCategory = generateChartDataByCategory(itemsWithCategory, TYPE_INCOME);

    let totalIncome = 0, totalOutcome = 0;
    itemsWithCategory.forEach(item => {
      if (item.category.type === TYPE_INCOME)
        totalIncome += item.price;
      if (item.category.type === TYPE_OUTCOME)
        totalOutcome += item.price;
    });

    return (
      // 空结点
      <React.Fragment>
        <header className="App-header">
          <div className="row">
            <div className="col-4">
              <MonthPicker
                year={currentDate.year}
                month={currentDate.month}
                onChange={this.changeDate}
              />
            </div>
            <div className="col-8">
              <TotalPrice
                income={totalIncome}
                outcome={totalOutcome}
              />
            </div>
          </div>
        </header>

        <div className="content-area py-3 px-3">
          {
            isLoading && <Loader />
          }

          {!isLoading &&
            <React.Fragment>
              <Tabs
                activeIndex={0}
                onTabChange={this.changeView}
              >
                <Tab>
                  <Ionicon
                    className="rounded-circle mr-2"
                    fontSize="25px"
                    color="#007bff"
                    icon='ios-paper'
                  />
              列表模式
            </Tab>
                <Tab>
                  <Ionicon
                    className="rounded-circle mr-2"
                    fontSize="25px"
                    color="#007bff"
                    icon='ios-pie'
                  />
              图表模式
            </Tab>
              </Tabs>
              <CreateBtn onCreateBtnClick={this.createItem} />
              {
                tabView === LIST_VIEW &&
                <PriceList
                  items={itemsWithCategory}
                  onModifyItem={this.modifyItem}
                  onDeleteItem={this.deleteItem}
                />
              }
              {
                tabView === CHART_VIEW &&
                <React.Fragment>
                  <PieChart title="本月支出" categoryData={chartOutcomeDateByCategory} />
                  <PieChart title="本月收入" categoryData={chartIncomeDateByCategory} />
                </React.Fragment>
              }
            </React.Fragment>
          }
        </div>
      </React.Fragment>
    );
  }
}

export default withRouter(withContext(Home));
