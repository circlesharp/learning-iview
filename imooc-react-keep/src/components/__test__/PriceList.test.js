import React from 'react';
import { shallow } from 'enzyme';
import Ionicon from 'react-ionicons';
import PriceList from '../PriceList';
import { items, categories } from '../../containers/Home';

const itemsWithCategory = items.map(item => {
  item.category = categories[item.cid];
  return item;
});

const props = {
  items: itemsWithCategory,
  onModifyItem: jest.fn(),
  onDeleteItem: jest.fn(),
};

let wrapper;

describe('PriceList component', () => {
  /* 每次运行单个测试用例时，都会执行 */
  beforeEach(() => {
    wrapper = shallow(<PriceList {...props} />);
  });

  /* 运行后，会在 __snapshots__ 出现 xx.test.js.snap */
  it('shoule render the component to match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  /* 测试条目是否等于 itemsWithCategory 的项的个数 */
  it('should render correct price items length', () => {
    expect(wrapper.find('.list-group-item').length).toEqual(itemsWithCategory.length);
  });

  /* 测试每项是否渲染正确的 icon&price */
  it('should render correct icon and price for each item', () => {
    const iconList = wrapper.find('.list-group-item').first().find(Ionicon);
    expect(iconList.length).toEqual(itemsWithCategory.length);
    expect(iconList.first().props().icon)
      .toEqual(itemsWithCategory[0].category.iconName);
  });

  /* 测试点击按钮是否触发特定回调 */
  it('should trigger the correct funtion callbacks', () => {
    const firstItem = wrapper.find('.list-group-item').first();

    // 模拟 修改 按钮被点击
    firstItem.find('.icon-modify').simulate('click');
    // 检查该函数是否被调用
    expect(props.onModifyItem).toHaveBeenCalled();

    // 检查该函数是否被调用且传入合适的参数
    firstItem.find('.icon-delete').simulate('click');
    expect(props.onDeleteItem).toHaveBeenCalledWith(itemsWithCategory[0]);
  });
});
