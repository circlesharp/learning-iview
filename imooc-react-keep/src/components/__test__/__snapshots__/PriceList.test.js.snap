// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PriceList component shoule render the component to match snapshot 1`] = `ShallowWrapper {}`;
